### PR TITLE
[RO-4] Asynchronous cache warmup

### PR DESCRIPTION
**Pull request by Skyld**
Ticket: https://oat-sa.atlassian.net/browse/RO-4

### Added
- Added [roster:cache-warmup:user](docs/cli/user-cache-warmer-command.md) command.
- Added [roster:cache-warmup:lti-instance](docs/cli/lti-instance-cache-warmer-command.md) command.
- Added possibility to profile CLI commands and HTTP calls with [Blackfire](docs/blackfire.md).

### Removed
- Removed `roster:cache:warmup` command.

**What has been changed on top of the list above:**

- Implemented Symfony Messenger integration for asynchronous cache warmup (optional).
- Infection MSI threshold has been increased from `95%` to `99%` in CI pipeline.
- PHPMD's Design/CouplingBetweenObjects rule has been enabled in CI pipeline (ticket: https://oat-sa.atlassian.net/browse/RO-72).
- Reworked cache warmup command documentations.
- Fixed various misconfigurations and composer dependencies in `prod` environment.
- Killed a ton of mutants in the code.

### CODE REVIEW

File: bin/console [hunk 1]
```diff
@@ -38,6 +42,22 @@ if ($_SERVER['APP_DEBUG']) {
     }
 }
 
-$kernel = new Kernel($_SERVER['APP_ENV'], (bool) $_SERVER['APP_DEBUG']);
+$isBlackfireProfilingRequested = $input->hasParameterOption('--blackfire', true);
```
Comment (andreluizmachado) [inline bin/console:45]: Suggestion: I'd suggest having an env variable o enable it Ex:

```
if ($_ENV['BLACKFIRE_ENABLED']) {
...
}

File: config/bundles.php [hunk 2]
```diff
@@ -13,7 +13,7 @@
     Doctrine\Bundle\FixturesBundle\DoctrineFixturesBundle::class => ['dev' => true, 'docker' => true, 'test' => true],
     Nelmio\Alice\Bridge\Symfony\NelmioAliceBundle::class => ['dev' => true, 'docker' => true, 'test' => true],
     Fidry\AliceDataFixtures\Bridge\Symfony\FidryAliceDataFixturesBundle::class => [
-        'dev' => true, 'docker' => true, 'test' => true,
+        'dev' => true, 'docker' => true, 'test' => true
```
Comment (andreluizmachado) [inline config/bundles.php:16]: is this change needed?

File: public/index.php [hunk 3]
```diff
@@ -51,6 +53,22 @@
 
 $kernel = new Kernel($_SERVER['APP_ENV'], (bool)$_SERVER['APP_DEBUG']);
 $request = Request::createFromGlobals();
+
+$isBlackfireProfilingRequested = $request->headers->has('X-Blackfire');
+
+if ($isBlackfireProfilingRequested) {
```
Comment (andreluizmachado) [inline public/index.php:59]: Suggestion: as the final user can manipulate the headers I'd suggest to have also an env variable o enable it Ex:

if ($isBlackfireProfilingRequested && $_ENV['BLACKFIRE_ENABLED']) {
...
}

File: src/Command/Ingester/AssignmentIngesterCommand.php [hunk 4]
```diff
@@ -67,7 +67,30 @@ protected function configure(): void
     {
         parent::configure();
 
-        $this->setDescription('Assignment data ingestion');
+        $this->setDescription('Ingests assignments into the application');
+        $this->setHelp(<<<'EOF'
+The <info>%command.name%</info> command ingests assignments into the application.
```
Comment (andreluizmachado) [inline src/Command/Ingester/AssignmentIngesterCommand.php:72]: Suggestion: Is it possible to have it in a file something like $this->setHelp(file_get_contents('file'))?
I believe that it'd improve code readability.

**Review by andreluizmachado — APPROVED** [review]
Hey @Skyld. Nice Job. Just a few comments:)

**Review by andreluizmachado — CHANGES_REQUESTED** [review]
Hey @Skyld. Nice Job. Just a few comments:)

File: bin/console (hunk 1 above)
Comment (Skyld) [reply bin/console:45]: Great idea! Added! :)

File: config/bundles.php (hunk 2 above)
Comment (Skyld) [reply config/bundles.php:16]: Not at all, it was just a merge problem, fixed!

File: public/index.php (hunk 3 above)
Comment (Skyld) [reply public/index.php:59]: Good suggestion, fixed!

File: src/Command/Ingester/AssignmentIngesterCommand.php (hunk 4 above)
Comment (Skyld) [reply src/Command/Ingester/AssignmentIngesterCommand.php:72]: Not a big fan of the idea to be honest. I think it's better to have the text here rather than in a separate file. But I admit it's not beautiful :/

File: .env [hunk 5]
```diff
@@ -48,6 +53,16 @@ LTI1P3_REGISTRATION_ID=demo
 ###< LTI ###
 
 ###> blackfire ###
-BLACKFIRE_SERVER_ID=<your_backfire_id>
-BLACKFIRE_SERVER_TOKEN=<your_backfire_secret>
+BLACKFIRE_ENABLED=false
```
Comment (luisfn) [inline .env:56]: Would be good to have some sort of presentation/training about blackfire in the future. 
Have no exposure to it at all

File: docs/blackfire.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
+# Profiling with Blackfire
```
Comment (luisfn) [inline docs/blackfire.md:1]: Happy to see this doc!

File: infection.json [hunk 7]
```diff
@@ -7,15 +7,33 @@
             "Kernel.php"
         ]
     },
-    "timeout": 10,
+    "timeout": 1000,
```
Comment (luisfn) [inline infection.json:10]: This won't increase too much the checking time for infection?

File: src/Lti/Model/Lti1p1LaunchUrl.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
+<?php
```
Comment (luisfn) [inline src/Lti/Model/Lti1p1LaunchUrl.php:1]: This class is almost the same as https://github.com/oat-sa/simple-roster/blob/develop/src/Lti/Request/LtiRequest.php (That in my opinion is not a `request`)

Can we have just one of them?

File: src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                $this->refreshCacheForUsername($username);
+            } catch (Throwable $exception) {
+                $errorLog = sprintf(
+                    "Unsuccessful cache warmup for user '%s'. Error: %s",
+                    $username,
+                    $exception->getMessage()
+                );
+
+                $this->messengerLogger->error($errorLog);
+                $this->cacheWarmupLogger->error($errorLog);
```
Comment (luisfn) [inline src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php:106]: Are you sure we want to replicate the same log into different channels?

File: src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        }
+
+        $logMessage = sprintf("Result cache for user '%s' was successfully warmed up", $username);
+        $logContext = [
+            'cacheKey' => $resultCacheId,
+            'cacheTtl' => $this->cacheTtl,
+        ];
+
+        $this->messengerLogger->info($logMessage, $logContext);
+        $this->cacheWarmupLogger->info($logMessage, $logContext);
```
Comment (luisfn) [inline src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php:144]: Same question as above

File: .env (hunk 5 above)
Comment (Skyld) [reply .env:56]: Good idea, I'll figure out something :)

File: src/Service/Cache/UserCacheWarmerService.php [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            $this->cacheWarmupLogger->info($log);
+        } catch (Throwable $exception) {
+            $errorLog = sprintf(
+                "Unsuccessful cache warmup for user '%s'. Error: %s",
+                implode(', ', $usernames),
+                $exception->getMessage()
+            );
+
+            $this->messengerLogger->error($errorLog);
+            $this->cacheWarmupLogger->error($errorLog);
```
Comment (luisfn) [inline src/Service/Cache/UserCacheWarmerService.php:98]: Same question as other files with this loggers

**Review by luisfn — COMMENTED** [review]
Great work Dani, it seems really good. Have just questions about it, but in overall it's good to go.

I have just concern when we add things out of the scope of the task like the LTI refactor here. IMO it brings some drawbacks like:
- Increase the review time
- Can interfere in parallel works

I know it's an improvement, but to avoid confusion I would suggest it to be done at least in separate PR.

File: src/Command/Ingester/AssignmentIngesterCommand.php (hunk 4 above)
Comment (andreluizmachado) [reply src/Command/Ingester/AssignmentIngesterCommand.php:72]: ok. not a problem for me, just a suggestion :)

**Review by andreluizmachado — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Changelog is updated according to changes (if applicable)
- [x] Documentation is updated according to changes (if applicable)

File: infection.json (hunk 7 above)
Comment (Skyld) [reply infection.json:10]: Actually this is a bug I think, in the documentation it says seconds but I'm 100% sure it's not in seconds. I was playing with the numbers and this was the magic number where I didn't get timeouts, and as you can see in the CI report it has no impact on the overall time (which is weird I admit)

File: src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php (hunk 9 above)
Comment (Skyld) [reply src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php:106]: Fixed based on our discussion

File: src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php (hunk 10 above)
Comment (Skyld) [reply src/MessageHandler/WarmUpGroupedUserCacheMessageHandler.php:144]: Fixed based on our discussion

File: src/Service/Cache/UserCacheWarmerService.php (hunk 11 above)
Comment (Skyld) [reply src/Service/Cache/UserCacheWarmerService.php:98]: Fixed based on our discussion

**Review by luisfn — APPROVED** [review]
**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)
